WebCore::Timer should assert that web thread lock is held or the web thread is disabled

rniwa opened the pull request:
#### a01578c7c53b262de50aa8dcc9923dfbe0185d15
<pre>
WebCore::Timer should assert that web thread lock is held or the web thread is disabled
<a href="https://bugs.webkit.org/show_bug.cgi?id=277299">https://bugs.webkit.org/show_bug.cgi?id=277299</a>

Reviewed by Wenson Hsieh.

Release assert WebThreadIsLockedOrDisabled in TimerBase::TimerBase and TimerBase::setNextFireTime.

Because WebCore::Timer always schedules timer in the web thread, it is incorrect to use this class
in the main thread of the UI process (which may or may not also use WebKitLegacy).

* Source/WebCore/platform/Timer.cpp:
(WebCore::TimerBase::TimerBase):
(WebCore::TimerBase::setNextFireTime):

Canonical link: <a href="https://commits.webkit.org/281614@main">https://commits.webkit.org/281614@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b18253a0443eea51147694f81ebc1025cdd925dd

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/60415 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/39766 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/12973 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/64358 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/10946 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/62545 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/47438 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/11179 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/48910 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/7618 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/62446 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/37051 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/52338 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/29751 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/33748 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/9564 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/9863 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/55637 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/9852 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/66066 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/4348 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/9693 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/56268 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/4367 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/52313 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/56430 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/3614 "Passed tests") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/9083 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/35576 "Built successfully") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/36658 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/37750 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/36402 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebCore/platform/Timer.cpp:
```diff
@@ -499,6 +506,9 @@ void TimerBase::updateHeapIfNeeded(MonotonicTime oldTime)
 void TimerBase::setNextFireTime(MonotonicTime newTime)
 {
+#if USE(WEB_THREAD)
+    ELEASE_ASSERT(WebThreadIsLockedOrDisabled());
```
`RELEASE_ASSERT`?

rniwa replied:
Oops, fixed.

ddkilzer commented on Source/WebCore/platform/Timer.cpp:
```diff
@@ -280,6 +284,9 @@ static_assert(sizeof(DeferrableOneShotTimer) == sizeof(SameSizeAsDeferrableOneSh
 {
+#if USE(WEB_THREAD)
+    ELEASE_ASSERT(WebThreadIsLockedOrDisabled());
+#endif
```
If I remember correctly, this check is too simplistic because:
- You must only perform this check in the UI process, not in any `com.apple.WebKit.*` processes.
- The timer is valid if the WebThread is disabled or if `(WebThreadIsCurrent() || WebThreadIsLocked())`, not just if `WebThreadIsLockedOrDisabled()`.

See this bug and the last PR posted:

Bug [185330](https://bugs.webkit.org/show_bug.cgi?id=185330): [iOS] Add assert to catch improper use of WebCore::Timer in UI Process
- Patch v5:  <https://bugs.webkit.org/attachment.cgi?id=344224&action=review>

As I recall, Patch v5 still wasn't quite correct (don't remember the details), and then I ran out of time to continue working on it.

whsieh replied:
FWIW, the web thread should _never_ be enabled in the web process in the first place (or any of the other WebKit child XPC services) — so the assertion would be relevant to the web process as well.

rniwa replied:
> FWIW, the web thread should _never_ be enabled in the web process in the first place (or any of the other WebKit child XPC services)

Right. `WebThreadIsLockedOrDisabled` should return true there because the web thread is never enabled in those processes.

rniwa replied:
> * You must only perform this check in the UI process, not in any `com.apple.WebKit.*` processes.
> * The timer is valid if the WebThread is disabled or if `(WebThreadIsCurrent() || WebThreadIsLocked())`, not just if `WebThreadIsLockedOrDisabled()`.

WebThreadIsLocked() returns true when WebThreadIsCurrent() is true.

ddkilzer reviewed (changes requested):
r- for build failure, and I think the check needs to only be active in the UI process, and it needs to allow for `WebThreadIsCurrent()` (even if the WebThread is not locked).

rniwa committed:
WebCore::Timer should assert that web thread lock is held or the web thread is disabled
https://bugs.webkit.org/show_bug.cgi?id=277299

Reviewed by Wenson Hsieh.

Release assert WebThreadIsLockedOrDisabled in TimerBase::TimerBase and TimerBase::setNextFireTime.

Because WebCore::Timer always schedules timer in the web thread, it is incorrect to use this class
in the main thread of the UI process (which may or may not also use WebKitLegacy).

* Source/WebCore/platform/Timer.cpp:
(WebCore::TimerBase::TimerBase):
(WebCore::TimerBase::setNextFireTime):

Canonical link: https://commits.webkit.org/281614@main
